ntoskrnl/ke/i386/cpu.c

```diff
@@ -1359,62 +1425,392 @@ KiCoprocessorError(VOID)
     __writecr0(__readcr0() | CR0_TS);
 }
 
-/*
- * @implemented
+/**
+ * @brief
+ * Saves the current floating point unit state
+ * context of the current calling thread. The
+ * saved floating unit state can be FX one
+ * if it's supported otherwise regular floating
+ * point state, FN in this case, is saved.
+ *
+ * @param[out] Save
+ * The saved floating point context given to the
+ * caller at the end of function's operations.
+ * The structure whose data contents are opaque
+ * to the calling thread.
+ *
+ * @return
+ * Returns STATUS_ILLEGAL_FLOAT_CONTEXT if the
+ * system doesn't support native FPU but it
+ * performs floating point software emulation
+ * instead. STATUS_INSUFFICIENT_RESOURCES is
+ * returned if pool allocation to store the FPU
+ * state has failed. STATUS_SUCCESS is returned
+ * indicating that the function has successfully
+ * completed its operations.
+ *
+ * @remarks
+ * Floating point context saving is done following
+ * this rule -- obtain the FPU state data from NPX
+ * registers or from the NPX frame based on NPX
+ * state itself if it's loaded or not. The function
+ * performs a FXSAVE or FNSAVE instruction depending
+ * on the system if it supports FX.
  */
 NTSTATUS
 NTAPI
-KeSaveFloatingPointState(OUT PKFLOATING_SAVE Save)
+KeSaveFloatingPointState(
+    _Out_ PKFLOATING_SAVE Save)
 {
-    PFNSAVE_FORMAT FpState;
+    PKPRCB CurrentPrcb;
+    PFX_SAVE_AREA FxSaveAreaFrame, FxSaveArea;
+    PFLOATING_SAVE_CONTEXT FsContext;
+    USHORT FsControlWord;
+    ULONG FsMXCsr, Cr0;
     ASSERT(KeGetCurrentIrql() <= DISPATCH_LEVEL);
-    UNIMPLEMENTED_ONCE;
 
-    /* check if we are doing software emulation */
+    /* Check if we are doing software emulation */
     if (!KeI386NpxPresent) return STATUS_ILLEGAL_FLOAT_CONTEXT;
 
-    FpState = ExAllocatePool(NonPagedPool, sizeof (FNSAVE_FORMAT));
-    if (!FpState) return STATUS_INSUFFICIENT_RESOURCES;
+    /* Cache the FS context */
+    FsContext = *((PVOID *) Save);
```

Comment:
Weird cast. `Save` is a pointer to a struct with perfectly defined layout, that doesn't correspond to the layout of any of the members of `FLOATING_SAVE_CONTEXT`.

Reply:
This is a trick used to initialize the floating point context from the parameter as a pointer that points to anything, as the members are opaque to the caller (one's not supposed to access them anyway). I don't think this would cause big issues down the road, but alternatively I can store the internal context by allocating a buffer for it with `ExAllocatePoolWithTag`, at the expense of taking up some bit of non paged memory though.